Highlight ROS Lyrical changes with examples (#6765)
ros2/ros2_documentation: source/Releases/Release-Lyrical-Luth.rst

Publish messages without copying data using ``rosidl::Buffer``
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Are you publishing data on ROS topics, but using the data elsewhere, like a GPU?
Tired of copying data out of the GPU before publishing just to copy it back into the GPU in the subscriber?
Use ``rosidl::Buffer`` to publish and subscribe ROS messages without moving data from elsewhere.

All ``uint8[]`` fields now have the type ``rosidl::Buffer<uint8_t>`` in C++ instead of ``std::vector<uint8_t>``.
Define your ROS messages with ``uint8[]`` fields and install an appropriate ``rosidl::BufferBackend`` implementation.
Note that only publishers and subscribers using ``rmw_fastrtps_cpp`` may use this feature for now, but `support in Zenoh is coming <https://github.com/ros2/rmw_zenoh/pull/930>`_.

Using a custom hardware accelerator or machine learning library?
You can benefit from this too.
See :doc:`../Tutorials/Advanced/Writing-a-Buffer-Backend` to learn how to implement your own ``rosidl::BufferBackend``.

> See :doc:`../Concepts/Intermediate/About-Buffer-Backends` and :doc:`../How-To-Guides/Using-Buffer-Backends` for more details.

Reviewer: @sloretz qq: This is NVIDIA's work right? Perhaps we want to point back to Discourse discussions?

> Use YAML tags in parameter files

Reviewer: ```suggestion
New Type Annotations in YAML Parameter Files
```

Perhaps a bit clearer?

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Tired of ``rcl`` interpreting ambiguous YAML parameter values as the wrong type?
Prevent that by specifying the type using YAML tags.

.. code-block:: yaml

  my_node:
    ros__parameters:
      string_param: !!str true
      bool_param: !!bool yes
      int_param: !!int 0
      float_param: !!float 10
      seq_param: !!seq [10, 0, -10]
      map_param: !!map {str: string, bool: true, int: 10, float: 1.1}

See `ros2/rcl#1275 <https://github.com/ros2/rcl/pull/1275>`_ for more info.

More logging options in launch files
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> Have different kinds of messages in your log files that you want to log with different severity?

Reviewer: ```suggestion
ROS Lyrical now supports per-message type log severity levels.
This feature lets you specify the log severity of a message based on the message's type. This makes it incredibly easy to bubble up specific messages in your log files for easier debugging!
```

        player = rosbag2_py.Player(storage_opts, play_opts)

        # Start the ROS node spinner and kick off the playback thread
        player.start_spin()
        player.play()

        # Verify playback startup and retrieve bag timing metadata
        print(player.wait_for_playback_to_start(1.0))
        print(player.wait_for_playback_to_start_exclusively(1.0))
        print(player.get_starting_time())
        print(player.get_playback_duration())

        # Control the playback state and step through messages manually
        print(player.is_paused())
        print(player.play_next())
        player.resume()
        player.pause()
        player.seek(0)

        # Await playback completion and terminate the worker threads
        print(player.wait_for_playback_to_finish(1.0))
        print(player.wait_for_playback_to_finish_exclusively(1.0))
        player.stop()
        player.stop_spin()

See `ros2/rosbag2#2047 <https://github.com/ros2/rosbag2/pull/2047>`_, `ros2/rosbag2#2062 <https://github.com/ros2/rosbag2/pull/2062>`_, `ros2/rosbag2#2061 <https://github.com/ros2/rosbag2/pull/2061>`_, and `ros2/rosbag2#2095 <https://github.com/ros2/rosbag2/pull/2095>`_ for more details.

> Circular recording by bag split

Reviewer: ```suggestion
Cap bag file storage use for disk constrained and long-running platforms
```

A little clearer perhaps.

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Recording data on your robot with limited disk space?
Try the new ``--max-bag-files`` option.
It limits the maximum number of bag files stored on disk by automatically deleting the oldest split files as new ones are created.

.. code-block:: bash

    # Max bag size: 100MB
    ros2 bag record --all --max-bag-size 100000000 --max-bag-files 5

See `ros2/rosbag2#2218 <https://github.com/ros2/rosbag2/pull/2218>`_ for more details.

More descriptive bag split names
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Do you have trouble identifying which bag is which?
``rosbag2`` now names bag splits so that each is self-descriptive and chronologically traceable.

.. code-block:: text

   {counter}_{prefix}_{timestamp}.{extension}

* **counter**: split index (integer starting from 0, *not zero-padded*)
* **prefix**: derived from the bag directory name, with any default timestamp suffix removed
* **timestamp**: local time at file creation, formatted as ``YYYY_MM_DD-HH_MM_SS``
* **extension**: bag file extension. e.g., ``.mcap``, ``.db3``

See `ros2/rosbag2#2265 <https://github.com/ros2/rosbag2/pull/2265>`_ for more details.

> ``rosbag2`` message-loss observability

Reviewer: ```suggestion
Catch data loss early with the new ``rosbag2`` message-loss observability
```

Control the publishing rate using ``--stats_max_publishing_rate``.

.. code-block:: bash

    # Publish message loss statistics at most 10 Hz
    ros2 bag record --all --stats_max_publishing_rate 10 -o /tmp/my_awesome_bag
    # If all is going well, you should see no output from this command
    ros2 topic echo /events/rosbag2_messages_lost

See `ros2/rosbag2#2039 <https://github.com/ros2/rosbag2/pull/2039>`_, `ros2/rosbag2#2144 <https://github.com/ros2/rosbag2/pull/2144>`_, and `ros2/rosbag2#2150 <https://github.com/ros2/rosbag2/pull/2150>`_ for more details.

``fish`` shell support
^^^^^^^^^^^^^^^^^^^^^^

Do you enjoy `fish shell <https://fishshell.com/>`_?
Do you want to use it with ROS?
Now you can!
Try out the new ``setup.fish`` script.

.. code-block:: shell

    source /opt/ros/lyrical/setup.fish

See `ros2/ros2cli#1211 <https://github.com/ros2/ros2cli/pull/1211>`_ and `ament/ament_package#164 <https://github.com/ament/ament_package/pull/164>`_ for more info.
To use ``fish`` shell with ``colcon``, check out `@Sunrisepeak's <https://github.com/Sunrisepeak>`_ `colcon-fish package <https://github.com/ros-x/colcon-fish>`_.

> ``ros2 param get`` a parameter from all nodes

Reviewer: Quick question, so this is effectively for global variables that are echoed locally on a node?

What happens if I have two nodes, each with a local `use_sim_time` parameter, where one is true and one is false? What does the output from that look like?

Author: > Quick question, so this is effectively for global variables that are echoed locally on a node?

In ROS 1, all parameters were "global" on a single parameter server. In ROS 2, all parameters are local to a node. If one node has `use_sim_time` to `True`, and another has `use_sim_time` set to False, then you'd see a node name followed by a the parameter value set on that node. I'll add a gif here so it's clearer.

See `ros2/ros2cli#1174 <https://github.com/ros2/ros2cli/pull/1174>`_ for more info.

``ros2 param`` get and set multiple parameters on one node
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Want to get and set multiple parameters on one node at the same time?
Use ``ros2 param get <node name> <param1> <param2> ...`` to get multiple values from a single node.

.. code-block:: console

    $ ros2 param get /robot_state_publisher frame_prefix ignore_timestamp publish_frequency
    frame_prefix:
      String value is:
    ignore_timestamp:
      Boolean value is: False
    publish_frequency:
      Double value is: 20.0

Use ``ros2 param set <node name> <param1> <value1> <param2> <value2> ...`` to set multiple values on a single node.

.. code-block:: console

    $ ros2 param set /robot_state_publisher frame_prefix foo ignore_timestamp True publish_frequency 10.0
    frame_prefix: Set parameter successful
    ignore_timestamp: Set parameter successful
    publish_frequency: Set parameter successful

See `ros2/ros2cli#1203 <https://github.com/ros2/ros2cli/pull/1203>`_ and `ros2/ros2cli#1204 <https://github.com/ros2/ros2cli/pull/1204>`_ for more details.

> Actions, Services, and Environment variables in ``ros2 doctor --report``

Reviewer: SIIIIIIIIIIIIIIIIIIIIIIICK!

> Actions, Services, and Environment variables in ``ros2 doctor --report``

Reviewer: ```suggestion
`ros2 doctor --report` now reports Actions, Services, and Environment variables 
```

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Most of the time in a ROS system the ``robot_state_publisher`` node does two things:

* It publishes the ``robot_description`` on a topic, and
* It publishes TF transforms given joint positions.

If you have ever tried to add ROS interfaces to a framework with its own internal robot model, you may have wished these were two separate utilities.
Now they can be!
Set the ``use_robot_description_topic`` parameter to ``true`` to make ``robot_state_publisher`` subscribe to the ``robot_description`` topic instead of publishing it.
Then, make the other robot framework publish its own robot description on that topic.

See `ros/robot_state_publisher#234 <https://github.com/ros/robot_state_publisher/pull/234>`_ for more info.

Resource retriever service
^^^^^^^^^^^^^^^^^^^^^^^^^^

Say you are debugging a robot in the field.
You open up RViz on your laptop, but you don't have the right version of the robot description installed.
In ROS Kilted, RViz added the ability to load meshes over the network using a ROS service ``/rviz/get_resource``; however, that ability was limited to RViz.

> ROS Lyrical comes with a generic ``resource_retriever_service`` so that any node can load meshes over the network.

Reviewer: Rad. This is the future.
